### PR TITLE
Add coverage with Coveralls and Blanket; moved tests to Mocha/Chai

### PR DESCRIPTION
**Pull request by joestump**
#### What does this PR do?

* Adds [Coveralls](https://coveralls.io/r/holidayextras/ui-toolkit) support via `npm run coveralls`.
* Switches test pipeline to Mocha/Chai instead of Jest/Jasmine.
* Coverage added via Blanket.

#### What unit or integration tests does this PR have?

n/a

#### What selenium tests does this PR have?

n/a

#### How should a developer review this?

You can run `npm run coveralls`, but it'll pollute the Coveralls UI with a build so make sure to delete any builds in Coveralls you create when manually testing. Travis is running successfully on this branch, which you can use to verify as well.

#### How should this be manually tested?

See above.

#### Any background context you want to provide?

We were using Jest, but the coverage with Istanbul was broken. After talking it over with @sambreed and the team, we decided to go with Mocha, which has wider tooling adoption and is less opinionated. 

#### What are the relevant tickets?

https://holidayextras.jira.com/browse/UXUI-253

#### Screenshots (if appropriate)

n/a

#### What gif best describes how you feel about this work?

---
#### Developer Definition of Done/Quality Checklist (for PR author to complete BEFORE code review):
- [x] I've checked there is appropriate unit, selenium and regression test coverage.
- [x] I've updated documentation with any changes to procedures.
- [x] I've tested this cross-browser/device for visual changes.
- [x] I've checked any significant visual changes with a member of the UX/UI pod.
- [x] I've checked this work against the requirements of the Jira.
- [x] I've manually checked this work against any dependent systems.
- [x] I've added and updated translations for all supported languages.
- [x] I've informed those who need to know of any unforeseen impact of this work (database migrations, deployment procedure changes).
- [x] I've checked that there are no negative speed or performance impacts from my work.
- [x] I've added appropriate tracking.
- [x] I am ready for this to be code reviewed, merged and tested.
- [x] I've thought about writing a blog post about this.

#### IP or Developer Review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [ ] +1 from me!

#### Software Engineer or Developer review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] +1 from me!

#### Software Engineer or project guru review:
- [x] I agree with the assumptions made in the quality checklist.
- [ ] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [ ] +1 from me!

#### Software Tester review
- [ ] I’ve run all the selenium tests locally and they pass.
- [ ] I agree with the test coverage.
- [ ] I've checked for Ruby changes and understand how to deploy if present.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,12 +13,18 @@
...
+    "scriptPreprocessor": "../preprocessor.js",
     "unmockedModulePathPatterns": [
       "./node_modules/"
     ],
     "testPathIgnorePatterns": [
       "./node_modules/"
+    ],
+    "testFileExtensions": [
+      "js",
+      "jsx"
```
Comment (joestump) [inline package.json]: @jackdcrawford if you modify `testFileExtensions` to have just `js` (or remove it entirely) it'll run coverage, but only show [coverage for a single file](https://coveralls.io/builds/2548358). If you run it with `jsx`, you get [coverage for everything](https://coveralls.io/builds/2548696) but Button and Flag tests fail.

File: README.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [![Build Status](https://travis-ci.org/holidayextras/ui-toolkit.svg)](https://travis-ci.org/holidayextras/ui-toolkit)
 [![Dependency Status](https://gemnasium.com/holidayextras/ui-toolkit.svg)](https://gemnasium.com/holidayextras/ui-toolkit)
+[![Coverage Status](https://coveralls.io/repos/holidayextras/ui-toolkit/badge.svg?branch=coveralls)](https://coveralls.io/r/holidayextras/ui-toolkit?branch=coveralls)
```
Comment (joestump) [inline README.md]: I'll remove this branch argument after it's merged to master.

**Comment by jackdcrawford** [comment]
I can take the SE/Dev review on this one

**Comment by hxoliverrumbelow** [comment]
I think the best word to describe some of the code in here is "funky" (in a good way).

**Comment by hxoliverrumbelow** [comment]
I haven't seen it working yet via Coveralls (I don't have a login) but everything else looks great, the tests run way faster too! I'll go find someone with a login and give this the :+1:

File: package.json [hunk 3]
```diff
@@ -9,17 +9,19 @@
   },
   "scripts": {
     "postinstall": "./scripts/postinstall.sh",
-    "test": "./node_modules/jest-cli/bin/jest.js",
-    "lint": "./node_modules/eslint/bin/eslint.js src/"
+    "coveralls": "./node_modules/mocha/bin/mocha --require blanket --require tests/bootstrap.js --reporter mocha-lcov-reporter --compilers .:tests/compiler.js src/**/__tests__/*-test.js src/components/**/__tests__/*-test.js | ./node_modules/coveralls/bin/coveralls.js",
+    "test": "./node_modules/mocha/bin/mocha --require tests/bootstrap.js --compilers .:tests/compiler.js src/**/__tests__/*-test.js src/components/**/__tests__/*-test.js",
+    "lint": "./node_modules/eslint/bin/eslint.js src/ && echo '🍻 All good!'"
```
Comment (hxoliverrumbelow) [inline package.json]: I've recently been pointed towards the documentation for npm-scripts which automatically adds `./node_modules/.bin/` to PATH before executing our custom commands: https://docs.npmjs.com/misc/scripts#path - with that in mind, could we drop all this prefixing?

**Comment by andrewcashmore** [comment]
Looks to be working - http://grab.by/HAyQ
Code looks good, just a few comments from @hxoliverrumbelow to address.

File: package.json (hunk 3 above)
Comment (joestump) [inline package.json]: I've been told this in the past as well. Sadly, it doesn't work. All commands fail locally for me when I remove the path. I'm going to leave this as-is given it works and changing it in the future won't modify how the commands are ran by developers.

**Comment by joestump** [comment]
This is now up to date with master and tests are passing locally. Should be good to merge once tests pass.

**Comment by jackdcrawford** [comment]
:+1:

**Pull request merged**